tools/linter/adapters/ufmt_linter.py

```diff
@@ -0,0 +1,247 @@
+    return LintMessage(
+        path=filename,
+        line=None,
+        char=None,
+        code="UFMT",
+        severity=LintSeverity.ERROR,
+        name="timeout",
+        original=None,
+        replacement=None,
+        description=(
+            "ufmt timed out while trying to process a file. "
+            "Please report an issue in pytorch/pytorch with the "
+            "label 'module: lint'"
+        ),
+    )
+
+
+def format_error_message(
+    filename: str, err: Union[OSError, subprocess.CalledProcessError]
+) -> LintMessage:
+    return LintMessage(
+        path=filename,
+        line=None,
+        char=None,
+        code="UFMT",
+        severity=LintSeverity.ADVICE,
+        name="command-failed",
+        original=None,
+        replacement=None,
+        description=(
+            f"Failed due to {err.__class__.__name__}:\n{err}"
+            if not isinstance(err, subprocess.CalledProcessError)
+            else (
+                "COMMAND (exit code {returncode})\n"
+                "{command}\n\n"
+                "STDERR\n{stderr}\n\n"
+                "STDOUT\n{stdout}"
+            ).format(
+                returncode=err.returncode,
+                command=" ".join(as_posix(x) for x in err.cmd),
+                stderr=err.stderr.decode("utf-8").strip() or "(empty)",
+                stdout=err.stdout.decode("utf-8").strip() or "(empty)",
+            )
+        ),
+    )
+
+
+def check_file(
+    filename: str,
+    retries: int,
+    timeout: int,
+) -> List[LintMessage]:
+    with open(filename, "rb") as f:
+        original = f.read()
+
+    try:
+        # Note that UFMT returns error code 1 if there are differences
+        # so the diff output is actually in the error
+        proc = run_command(
+            [sys.executable, "-mufmt", "format", filename],
+            input=None,
+            retries=retries,
+            timeout=timeout,
+        )
+
+        # UFMT is a bit unyielding to use in lintrunner workflow because it
+        # only includes options to check, generate diff, and apply the patch
+        # in place
+        with open(filename, "rb") as f:
+            replacement = f.read()
+
+        if original == replacement:
+            return []
+
+        return [
+            LintMessage(
+                path=filename,
+                line=None,
+                char=None,
+                code="UFMT",
+                severity=LintSeverity.WARNING,
+                name="format",
+                original=original.decode("utf-8"),
+                replacement=replacement.decode("utf-8"),
+                description="Run `lintrunner -a` to apply this patch.",
+            )
+        ]
+    except subprocess.TimeoutExpired:
+        return [format_timeout_message(filename)]
+    except (OSError, subprocess.CalledProcessError) as err:
+        return [format_error_message(filename, err)]
+    finally:
+        if not replacement or original != replacement:
```

Comment:
Instead of modifying the file system in-place and restoring it, can we just run usort then black back to back?

That's pretty much what ufmt is doing anyway https://github.com/omnilib/ufmt/blob/eecbc760c0e9489152a6ba9f073d1edb14c0b502/ufmt/core.py#L82-L92

Reply:
Sure, let me try that. I agree that it seems a bit awkward to fit the fire-and-forge formatting approach of `ufmt` into `lintrunner / lintrunner -a` workflow. Another workaround is to get the diff from ufmt and apply it as a patch to get the replacement text, i.e. `patch -u filename -o filename.patched`. However, I think that differs very little than the way the file is modified in place and restored here.

I guess the best approach is to submit a simple PR to ufmt to print the output to stdout instead of modifying the file in place, i.e. `ufmt --dry-run format filename`. If we have that, we can use ufmt out of the box instead of duplicating their code running usort and black back to back

Reply:
Awesome idea!

Reply:
In the meantime, I think I have managed to hide most of the complexity behind the `ufmt_string` function call. Thanks for pointing out the `ufmt` API to me. It seems like a reasonable approach here in which we can both reuse `ufmt` logic, and avoid its current handicap in term of the behavior of the CLI